### PR TITLE
feat: introduce notification system

### PR DESCRIPTION
**Pull request by alexcarpenter**
🔍 [Preview Link](https://react-components-git-acnotifications-system-hashicorp.vercel.app/components/notification)

---

## Description

This PR introduces our own custom notification system vs using react-hot-toast directly. This sets us up to implement differences in interactive and non-interactive toasts that is still WIP.

This PR also removes the toast package and moves the notification system functionality into the notification component to make it easier to make releases instead of having the separate packages. The notification system code lives within the core folder of the notifications component.

Since we're controlling the rendering of the toast, I also introduced Framer-Motion to handle the enter and exit animations.

### Usage

1. Add `<Notifications />` within your `_app.tsx` file. This is where you notification will be rendered from. Example placement below:

```jsx
import MotionConfig from '@hashicorp/react-motion-config'
import { Notifications } from '@hashicorp/react-notification`

function App({ Component, pageProps, layoutData }) {
  return (
     <MotionConfig>
      <Layout {...(layoutData && { data: layoutData })}>
        <Component {...pageProps} {...layoutData} />
      </Layout>
      {renderConsentManager && <ConsentManager />}
      <Notifications />
    </MotionConfig>
  )
}
```

2. Render notifications.

```tsx
import {
  notification,
  NotificationWithThumbnail,
} from "@hashicorp/react-notiication";

function example() {
  return (
    <button
      onClick={() =>
        notification((n) => {
          return (
            <NotificationWithThumbnail
              thumbnail={{
                src: "https://www.datocms-assets.com/2885/1651495958-infrastructure.png",
                alt: "Notfication thumbnail",
              }}
              description="lAliquet in lacus, scelerisque tristique purus. Lob ortis felis orci lacus egestas et turpis."
              cta={{
                title: "Standalone link",
                url: "/",
              }}
              onDismiss={() => notification.remove(n.id)}
            />
          );
        })
      }
    >
      Notification with thumbnail
    </button>
  );
}
```

Example usage can be found in the Localized Microsite redirects PR https://github.com/hashicorp/hashicorp-www-next/pull/2880

Follow up task to bump Framer-Motion to latest release https://app.asana.com/0/1187888078520425/1202608962158074

### PR Checklist 🚀

Items in this checklist may not may not apply to your PR, but please consider each item carefully.

- [x] Add Asana and Preview links above.
- [x] Conduct thorough self-review.
- [ ] Add or update tests as appropriate.
- [x] Conduct reasonable cross browser testing for both compatibility and responsive behavior (We have a [Sauce Labs](https://app.saucelabs.com/) account for this, if you don't have access, just ask!).
- [x] Conduct reasonable accessibility review (use the [WAS](https://accessible.org/Web-Accessibility-Standards-WAS-2.pdf) as a guide or an [axe browser plugin](https://www.deque.com/axe/) until we establish more formal checks).
- [ ] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

### CODE REVIEW

File: packages/notification/components/notifications/index.tsx
Comment (alexcarpenter) [reply packages/notification/components/notifications/index.tsx:7]: Yes, good call! https://github.com/hashicorp/react-components/pull/609/commits/85c50f73b4eba2cfadb8e6cc42c5bdb93685b763

File: packages/notification/core/store.ts
Comment (alexcarpenter) [reply packages/notification/core/store.ts:109]: This was my mistake in not updating the variable. We're sticking to using Notification naming vs Toast with the idea that we'd not be rendering a Toast with interactive content but a non-modal dialog.

File: packages/notification/components/notifications/index.tsx
Comment (alexcarpenter) [reply packages/notification/components/notifications/index.tsx:14]: https://user-images.githubusercontent.com/825855/173401867-071a6220-736f-4f69-a343-e6d1195eac62.mp4

File: packages/notification/core/notification.ts
Comment (alexcarpenter) [reply packages/notification/core/notification.ts:34]: This setup allows us to create methods on `notification` to do things like remove and dismiss notifications.

Below you can see how this is being used. 

```diff
notification((n) => {
  return (
    <Notification
      description="lAliquet in lacus, scelerisque tristique purus. Lob ortis felis orci lacus egestas et turpis."
      cta={{
        title: 'Standalone link',
        url: '/'
      }}
+      onDismiss={() => notification.remove(n.id)}
    />
  )
})
```

In the future when we offer the different notification types. I can envision an api for generating the type of notification via these methods.

```js
// Generate a notification as a toast
notfication.toast((n) => {})

// Generate a notification as a non-modal dialog
notfication.dialog((n) => {})
```

**Comment by BRKalow** [comment]
Can we put a link back to `react-hot-toast` if this is largely a derivative work? Just as a helpful breadcrumb and reference for future contributors! A reference in `packages/notifications/README.md` would be great.

File: packages/notification/core/notification.ts
Comment (BRKalow) [reply packages/notification/core/notification.ts:34]: ⛏️ nit: Ah it looks like in `react-hot-toast` `createHandler` accepts a type: https://github.com/timolins/react-hot-toast/blob/main/src/core/toast.ts#L35. Are we thinking we would bring back that capability to enable different types of notifications at this level? Otherwise if we're only using `createHandler` once like this I agree with Ashlee it might be clearer if `createHandler` was inlined into the definition of `notification`

File: packages/notification/package.json
Comment (BRKalow) [inline packages/notification/package.json:12]: So `framer-motion` kinda defacto becomes a dependency in the critical path as a result of this, along with its layout animation feature. I'm not sure the best way to handle this, but something to keep in mind. As soon as we start using this package in dev-portal our bundle will go up by like `~40kb`. Maybe we could try to use `m` instead of `motion` and use [LazyMotion](https://www.framer.com/docs/lazy-motion/) to include the features we need specifically.

File: packages/notification/core/notification.ts
Comment (alexcarpenter) [reply packages/notification/core/notification.ts:34]: @BRKalow yes, I simplified for now, but I did imagine once we understand a bit more how we'd introduce the different notification needs, that we'd have notification types "toast" and "dialog" which would then be treated accordingly when rendering.

File: packages/notification/core/notification.ts
Comment (BRKalow) [inline packages/notification/core/notification.ts:32]: I wonder if it would be possible for us to enforce usage of our `Notification*` components somehow, if that's something we want to do 🤔

**Comment by BRKalow** [comment]
Visually and functionally this looks good. I wonder if the API here is potentially too open-ended? In theory we can render anything using `notification()`, which I don't think is what we want in practice.

I believe we had some discussions about the API before and landed on where we are, but my memory is fuzzy 🤔

**Comment by BRKalow** [comment]
Visually and functionally this looks good. I wonder if the API here is potentially too open-ended? In theory we can render anything using `notification()`, which I don't think is what we want in practice.

I believe we had some discussions about the API before and landed on where we are, but my memory is fuzzy 🤔

**Review by BRKalow — COMMENTED** [review]
It would be great to see some tests of the `notification` method here along with the provider, in addition to some docs comments above exported methods / components!

File: packages/notification/components/notifications/index.tsx
Comment (ashleemboyer) [reply packages/notification/components/notifications/index.tsx:14]: Nice! Would love to see that visual indicator become a thing in a future iteration. 😊

File: packages/notification/core/store.ts
Comment (ashleemboyer) [reply packages/notification/core/store.ts]: Sounds good to me!

File: package.json
Comment (alexcarpenter) [inline package.json:34]: Needed for testing the `useNotification` hook.

File: packages/notification/package.json
Comment (alexcarpenter) [reply packages/notification/package.json:13]: Introduced [`MotionConfig`](https://github.com/hashicorp/react-components/tree/main/packages/motion-config) component to manage lazy loading framer motions features.

File: next.config.js
Comment (BRKalow) [inline next.config.js:55]: This should be included in `v2.1.0` of `@hashicorp/platform-nextjs-plugin`

**Review by BRKalow — APPROVED** [review]
🏆

File: next.config.js
Comment (dstaley) [reply next.config.js:55]: @BRKalow yeah we need another release to handle the `react-motion-config -> framer-motion` resolution chain. This was a temporary workaround to unblock swingset previews :)

Comment (dstaley) [reply next.config.js:55]: https://github.com/hashicorp/web-platform-packages/pull/66

**Pull request merged**
